File: scrimage-core/src/main/scala/com/sksamuel/scrimage/Position.scala
/*
   Copyright 2013 Stephen K Samuel

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 */

package com.sksamuel.scrimage

/** @author Stephen Samuel */
sealed trait Position {

  /**
   * Returns the x coordinate for where a target should be placed inside the source.
   */
  def calculateX(sourceWidth: Int, sourceHeight: Int, targetW: Int, targetH: Int): Int

  /**
   * Returns the y coordinate for where an image should be placed inside the canvas.
   */
  def calculateY(sourceWidth: Int, sourceHeight: Int, targetW: Int, targetH: Int): Int

  def calculateXY(sourceWidth: Int, sourceHeight: Int, targetW: Int, targetH: Int): (Int, Int) =
    (calculateX(sourceWidth, sourceHeight, targetW, targetH),
      calculateY(sourceWidth, sourceHeight, targetW, targetH))
}

trait Left extends Position {
  def calculateX(sourceWidth: Int, sourceHeight: Int, targetW: Int, targetH: Int): Int = 0
}

trait Right extends Position {
  def calculateX(sourceWidth: Int, sourceHeight: Int, targetW: Int, targetH: Int): Int = sourceWidth - targetW
}

trait Top extends Position {
  def calculateY(sourceWidth: Int, sourceHeight: Int, targetW: Int, targetH: Int): Int = 0
}

trait Bottom extends Position {
  def calculateY(sourceWidth: Int,
                 sourceHeight: Int,
                 targetW: Int,
                 targetH: Int): Int = sourceHeight - targetH
}

trait CenterX extends Position {
  def calculateX(sourceWidth: Int, sourceHeight: Int, targetW: Int, targetH: Int): Int =
    ((sourceWidth - targetW) / 2.0).toInt
}

trait CenterY extends Position {
  def calculateY(sourceWidth: Int, sourceHeight: Int, targetW: Int, targetH: Int): Int =
    ((sourceHeight - targetH) / 2.0).toInt
}

object Position {
  object TopLeft extends Top with Left
  object TopCenter extends Top with CenterX
  object TopRight extends Top with Right
  object CenterLeft extends CenterY with Left
  object Center extends CenterY with CenterX
  object CenterRight extends CenterY with Right
  object BottomLeft extends Bottom with Left
  object BottomCenter extends Bottom with CenterX
  object BottomRight extends Bottom with Right
}

File: scrimage-core/src/main/java/com/sksamuel/scrimage/Position.java
package com.sksamuel.scrimage;

public interface Position {

    class _TopCenter implements Top, CenterX {}
    class _TopLeft implements Top, Left {}
    class _TopRight implements Top, Right {}
    class _CenterLeft implements CenterY, Left {}
    class _Center implements CenterY, CenterX {}
    class _CenterRight implements CenterY, Right {}
    class _BottomLeft implements Bottom, Left {}
    class _BottomCenter implements Bottom, CenterX {}
    class _BottomRight implements Bottom, Right {}

    Position TopCenter = new _TopCenter();
    Position TopLeft = new _TopLeft();
    Position TopRight = new _TopRight();
    Position CenterLeft = new _CenterLeft();
    Position Center = new _Center();
    Position CenterRight = new _CenterRight();
    Position BottomCenter = new _BottomCenter();
    Position BottomRight = new _BottomRight();
    Position BottomLeft = new _BottomLeft();

    /**
     * Returns the x coordinate for where a target should be placed inside the source.
     */
    int calculateX(int sourceWidth, int sourceHeight, int targetWidth, int targetHeight);

    /**
     * Returns the y coordinate for where an image should be placed inside the canvas.
     */
    int calculateY(int sourceWidth, int sourceHeight, int targetWidth, int targetHeight);

    default Dimension calculateXY(int sourceWidth, int sourceHeight, int targetWidth, int targetHeight) {
        return new Dimension(
                calculateX(sourceWidth, sourceHeight, targetWidth, targetHeight),
                calculateY(sourceWidth, sourceHeight, targetWidth, targetHeight)
        );
    }
}

interface Left extends Position {
    default int calculateX(int sourceWidth, int sourceHeight, int targetWidth, int targetHeight) {
        return 0;
    }
}

interface CenterX extends Position {
    default int calculateX(int sourceWidth, int sourceHeight, int targetWidth, int targetHeight) {
        return (int) ((sourceWidth - targetWidth) / 2.0);
    }
}

interface Right extends Position {
    default int calculateX(int sourceWidth, int sourceHeight, int targetWidth, int targetHeight) {
        return sourceWidth - targetWidth;
    }
}

interface Top extends Position {
    default int calculateY(int sourceWidth, int sourceHeight, int targetWidth, int targetHeight) {
        return 0;
    }
}

interface CenterY extends Position {
    default int calculateY(int sourceWidth, int sourceHeight, int targetWidth, int targetHeight) {
        return (int) ((sourceHeight - targetHeight) / 2.0);
    }
}

interface Bottom extends Position {
    default int calculateY(int sourceWidth,
                           int sourceHeight,
                           int targetWidth,
                           int targetHeight) {
        return sourceHeight - targetHeight;
    }
}

